src/core/ts/background-scripts/background.js/lib/BergamotTranslatorAPI.ts

```diff
@@ -256,58 +312,74 @@ class TranslationRequestManager {
     const { loadModelParams } = translateParams;
     const { from, to } = loadModelParams;
     const languagePair = `${from}${to}`;
-    Telemetry.global.record(() => fromLang.set(from), "fromLang");
-    Telemetry.global.record(() => toLang.set(to), "toLang");
 
     // First check if we need to load a model
     if (!this.loadedLanguagePair || this.loadedLanguagePair !== languagePair) {
-      const loadModelStart = performance.now();
-
+      const modelWillLoadEventData: ModelWillLoadEventData = {
+        requestId,
+        loadModelParams,
+      };
+      this.dispatchEvent(
+        new CustomEvent("modelWillLoad", {
+          detail: modelWillLoadEventData,
+        }),
+      );
       const loadModelRequestWorkerMessage: LoadModelRequestWorkerMessage = {
         type: "loadModel",
         requestId,
         loadModelParams,
       };
-      await workerManager.loadModel(loadModelRequestWorkerMessage);
+      const loadModelResults = await workerManager.loadModel(
+        loadModelRequestWorkerMessage,
+      );
       this.loadedLanguagePair = languagePair;
-
-      const loadModelEnd = performance.now();
-      // todo: replace to timespan when it is supported
-      Telemetry.global.record(
-        () => modelLoadTime.set(String(loadModelEnd - loadModelStart)),
-        "loadModelTime",
+      const modelLoadedEventData: ModelLoadedEventData = {
+        requestId,
+        loadModelParams,
+        loadModelResults,
+      };
+      this.dispatchEvent(
+        new CustomEvent("modelLoaded", {
+          detail: modelLoadedEventData,
+        }),
       );
     }
 
     // Send the translation request
     const start = performance.now();
-
     const translationResults = await workerManager.translate(
       translateRequestWorkerMessage,
     );
-    this.queuedRequestsByRequestId.get(requestId).resolve(translationResults);
 
+    // Summarize performance stats
     const end = performance.now();
-    // todo: replace to timespan when it is supported
-    const timeSpentMs = end - start;
-    Telemetry.global.record(
-      () => translationTime.set(String(timeSpentMs)),
-      "translateTime",
+    const translationWallTimeMs = end - start;
+    const originalTextsTranslationPerformanceStats = translationPerformanceStats(
+      translationResults.originalTexts,
+      translationWallTimeMs,
     );
-
-    // we might want to pass this info from wasm if it exists
-    const wordsNumber = translateParams.texts
-      .map(x => x.split(" ").length)
-      .reduce((a, b) => a + b, 0);
-    const speed = Math.floor((wordsNumber / timeSpentMs) * 1000);
-    // todo: replace to quantity when it is supported
-    Telemetry.global.record(
-      () => wordsPerSecond.set(String(speed)),
-      "translateSpeed",
+    const translatedTextsTranslationPerformanceStats = translationPerformanceStats(
```

Comment:
We should decide what "words" in "words per second" corresponds to. Eg is it always the source language word count? Always the target language word count? Or always the English language word count? Different languages for sure have different average word lengths, affecting the numbers. Maybe this should be decided in the scope of bergamot-evaluation, after which we can revisit what we want to measure in telemetry?

Reply:
Let's decide that this is the word count of the source text, doesn't matter what language it is in. We work with languages from a similar language group, no so this will work for us. I wouldn't be concerned about language differences at this point. We can also add characters per second later if it's not enough. It is a user-centric metric that shows in how much time how many words displayed on a user's screen are translated. Also "source tokens per second" is a standard metric to measure speed in Marian paper for example https://www.aclweb.org/anthology/P18-4020.pdf, so let's just stick with it. The problem here is that we don't have access to Marian tokenizer and we approximate it with our naive splitting, but it's fine for telemetry I think.

Reply:
Sure. Decided! This is also what is being included in telemetry currently, so that is good. Want me to remove the attribute in this thread's codeline?

Reply:
yep, let's remove it then

Reply:
Done

Reply:
The metric in that paper is literally wc -w without any fancy tokenization, so just count spaces?